keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,807 @@
+- Basic unit tests
+
+#### Alpha -> Beta Graduation
+
+- API server support for `webhook` `AuthenticationConfig` type
+- Kubelet support for `x509` `AuthenticationConfig` type
+- Complete server side validation of `AuthenticationConfig` REST API
+- Complete unit and integration tests
+
+#### Beta -> GA Graduation
+
+- 3 examples of real world usage
+- Gather feedback from example usages to confirm API structure is sufficient
+- Associated documentation
+- End-to-end test that deploys dex on cluster
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to keep previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to make use of the enhancement?
+-->
+
+No actions are required on upgrade as the `AuthenticationConfig` API is inert
+until used.
+
+If a cluster is downgraded to a version that does not support the
+`AuthenticationConfig` API, all objects will become inert.  Thus care should be
+taken before downgrading to ensure that some form of authentication will still
+be available to the user after the downgrade is complete.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+An n-2 kubelet may not support x509 authentication via the `AuthenticationConfig`
+API.  This can be mitigated by explicitly configuring the kubelet via the
+`--client-ca-file` flag.
+
+## Implementation History
+
+<!--
+Major milestones in the life cycle of a KEP should be tracked in this section.
+Major milestones might include
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+2020-04-15: Initial KEP draft created
+2020-05-15: KEP updated to address various comments
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+TBD.
+
+## Alternatives
```

Comment:
Responding to Jordan's overall comment about "make OIDC better."

Overall, I do not understand how a REST API (I discuss two below) ends up being different than the impersonation API or the CSR API from a security perspective.  If we give the cluster operator the same level of control, why does it matter?

As a thought experiment, if I was to propose a new `UserTokenRequestAPI` that worked like the CSR API's `kubernetes.io/kube-apiserver-client` signer, would that be okay?  This would be a way for some privileged component to ask for a _token_ for a particular user info that the Kube API would honor.  The token itself would be opaque to the client.  What ever component issued the token could choose not to based on whatever requirements the cluster-operator specifies.  This handles the "the identity is rooted in a trust owned the API server / cluster operator."  Such an API adds one layer of indirection but that could be hidden from the end user.

---

> My recommendation would be to narrow this to improving the OIDC provider config, and for clusters that wanted to allow control of that config via a REST API, to use a CRD-based API paired with a config file writer.
 
>     1. allow the kube-apiserver to configure multiple oidc providers via a structured, reloadable config file
> 
>     2. Separately, create a CRD for custom resources that map to the items in the oidc config in a straightforward way
> 
>     3. Providers that want to expose the config via a REST API can install the CRD and the custom-resource -> OIDC config file process
> 
>     4. Providers that do not want to expose the config still benefit from multiple oidc provider capability and have no additional exposure and no new API surface to lock down

So we would need to in `k/k`:

1. Make a new file based API
1. Wire that file based config through
1. Reload that config

Outside of `k/k`:

1. Create a new CRD
1. Create new controllers to handle the CRD
1. Create new admission plugins to restrict access to the CRD, but with far weaker security (they can be removed if you can delete the webhook)

For every single provider:

1. Ask them to run all of those components co-located with their API servers
1. Configure credentials and the correct authz for these components
1. Set the right flags on the API server

This sounds an incredible amount of work in-tree, out-of-tree, and for every single provider.  I understand the desire to build things out of core, but I do not think this moves us forward in any meaningful way.

Why would we instead not:

In `k/k`:

1. Create a new dynamic OIDC config API, in a new group, disabled by default
1. Wire the new API through
1. Reload the API config

Outside of `k/k`:

No work

For every single provider:

1. Ask them to pass a single flag to enable the API if they wish to support it

This is the same amount of work and changes in `k/k`.  There is no work outside of `k/k` and the burden on the providers is as small as we can make it.  Since its off by default, no security concerns are added.  The provider would opt-in in the same way they would for the CRD based flow.  They just would not have to do a large amount of provider specific work.

---

> Is there evidence this is a thing most hosted Kubernetes offerings want and plan to enable? If this would not be enabled by default, or made part of conformance, or opted into broadly, I question whether it should exist as a built-in API at all.

From what I can tell, EKS and AKS would.  GKE and OpenShift would not (though OpenShift already lets you control IDPs via a Kube REST API).  Not sure about others.  I suspect Linode, Digital Ocean, Rancher might.  It would be enabled in VMware's products (duh).

If every hosted offering had a Kube API that let you configure auth in some way that they were comfortable with, then we would be fine as-is.  The reality is that even when providers want to allow this, it is a lot of work to build all the tooling to wire this stuff through.  They also tend to use non Kube APIs, which makes it hard to integrate with.